hbase-server/src/main/java/org/apache/hadoop/hbase/wal/WALTailingReader.java

```diff
@@ -0,0 +1,148 @@
+     * the compression context when reseting, which means when calling resetTo method, we need to
+     * skip to the position instead of just seek to, which will impact performance.
+     */
+    ERROR_AND_RESET_COMPRESSION,
+    /**
+     * This means we reach the EOF and the upper layer need to reset to see if there is more data.
+     * Notice that this does not mean that there is necessarily more data, the upper layer should
+     * determine whether they need to reset and read again.
+     */
+    EOF_AND_RESET,
+    /**
+     * Mostly the same with the above {@link #EOF_AND_RESET}, the difference is that here we also
+     * mess up the compression dictionary when reading data, so the upper layer should also clear
+     * the compression context when reseting, which means when calling resetTo method, we need to
+     * skip to the position instead of just seek to, which will impact performance. The
+     * implementation should try its best to not fall into this situation.
+     */
+    EOF_AND_RESET_COMPRESSION;
+
+    /**
+     * A dummy result for returning, as except {@link NORMAL}, for other state we do not need to
+     * provide fields other than state in the returned {@link Result}.
+     */
+    private Result result = new Result(this, null, -1);
+
+    public Result getResult() {
+      return result;
+    }
+
+    public boolean resetCompression() {
+      return this == ERROR_AND_RESET_COMPRESSION || this == EOF_AND_RESET_COMPRESSION;
+    }
+
+    public boolean eof() {
+      return this == EOF_AND_RESET || this == EOF_AND_RESET_COMPRESSION || this == EOF_WITH_TRAILER;
+    }
+  }
+
+  final class Result {
+
+    private final State state;
+    private final Entry entry;
+    private final long entryEndPos;
+
+    public Result(State state, Entry entry, long entryEndPos) {
+      this.state = state;
+      this.entry = entry;
+      this.entryEndPos = entryEndPos;
+    }
+
+    public State getState() {
+      return state;
+    }
+
+    public Entry getEntry() {
+      return entry;
+    }
+
+    public long getEntryEndPos() {
+      return entryEndPos;
+    }
+  }
+
+  /**
+   * Read the next entry and make sure the position after reading does not go beyond the given
+   * {@code limit}.
+   * <p/>
+   * Notice that we will not throw any checked exception out, all the states are represented by the
+   * return value. Of course we will log the exceptions out. The reason why we do this is that, for
+   * tailing a WAL file which is currently being written, we will hit EOFException many times, so it
+   * should not be considered as an 'exception' and also, creating an Exception is a bit expensive.
+   * @param limit the position limit. See HBASE-14004 for more details about why we need this
+   *              limitation. -1 means no limit.
+   */
+  Result next(long limit);
+
+  /**
+   * Get the current reading position.
+   */
+  long getPosition() throws IOException;
```

Comment:
I noticed that `WALTailingReader` and `WALStreamReader` these two interfaces both declare this method, totally same. Then the specific implementation classes `ProtobufWALStreamReader` and `ProtobufWALTailingReader` do not implement it, but implements it through the parent abstract class `AbstractProtobufWALReader`.
So I think maybe you can delete it in these two interfaces.

Reply:
The upper layer can only get a WALTailingReader or WALStreamReader, they can not see the AbstractProtobufWALReader... So if we remove this method from these two interfaces, the upper layer classes can not use this method any more...

Reply:
Got it. Thank you.

I've read through the entire patch as best I can, and haven't found anything wrong.